File: backend/src/inf_canvas/schema/canvas.py
"""Canvas data model (Pydantic mirror of canvas.ts)."""

from datetime import UTC, datetime
from typing import Any, Literal

from pydantic import BaseModel

from .equipment import EquipmentType

LineType = Literal["process", "signal", "electrical", "pneumatic"]


class Position(BaseModel):
    x: float
    y: float


class CanvasNode(BaseModel):
    id: str
    type: EquipmentType
    position: Position
    label: str | None = None
    rotation: float | None = None
    data: dict[str, Any] | None = None


class PipeData(BaseModel):
    lineType: LineType | None = None
    label: str | None = None
    animated: bool | None = None


class CanvasEdge(BaseModel):
    id: str
    source: str
    sourcePort: str
    target: str
    targetPort: str
    data: PipeData | None = None


class CanvasMeta(BaseModel):
    id: str
    name: str
    createdAt: str | None = None
    updatedAt: str | None = None


class CanvasState(BaseModel):
    meta: CanvasMeta
    nodes: list[CanvasNode] = []
    edges: list[CanvasEdge] = []


def empty_canvas(canvas_id: str, name: str) -> CanvasState:
    now = datetime.now(UTC).isoformat()
    return CanvasState(meta=CanvasMeta(id=canvas_id, name=name, createdAt=now, updatedAt=now))

File: frontend/src/schema/canvas.ts
/** Authoritative canvas data model. Persisted as JSON; shared by both sides. */

import type { EquipmentType } from './equipment';

export interface Position {
  x: number;
  y: number;
}

export interface CanvasNode {
  id: string;
  type: EquipmentType;
  position: Position;
  /** Tag / display label, e.g. "P-101". */
  label?: string;
  /** Rotation in degrees (0, 90, 180, 270). */
  rotation?: number;
  /** Free-form attributes (tag number, design specs, source bbox, ...). */
  data?: Record<string, unknown>;
}

export type LineType = 'process' | 'signal' | 'electrical' | 'pneumatic';

export interface PipeData {
  lineType?: LineType;
  /** Line number / pipe spec, e.g. '6"-P-1001-CS'. */
  label?: string;
  animated?: boolean;
  /** Manual routing points the pipe passes through (reroute). */
  waypoints?: Position[];
}

export interface CanvasEdge {
  id: string;
  source: string;
  sourcePort: string;
  target: string;
  targetPort: string;
  data?: PipeData;
}

/** A titled background region grouping equipment (e.g. by unit/system). */
export interface CanvasGroup {
  id: string;
  label: string;
  position: Position;
  width: number;
  height: number;
  color?: string;
  /** Node ids that move with the group. */
  memberIds: string[];
}

export interface CanvasMeta {
  id: string;
  name: string;
  /** ISO-8601 timestamps. */
  createdAt?: string;
  updatedAt?: string;
}

export interface CanvasState {
  meta: CanvasMeta;
  nodes: CanvasNode[];
  edges: CanvasEdge[];
  groups: CanvasGroup[];
}

export function emptyCanvas(id: string, name: string): CanvasState {
  return {
    meta: { id, name },
    nodes: [],
    edges: [],
    groups: [],
  };
}
